lib/common/xml.c

```diff
@@ -1860,79 +1997,32 @@ replace_node(xmlNode *old, xmlNode *new)
  * \return Standard Pacemaker return code
  */
 int
-pcmk__xe_find_replace(xmlNode *parent, xmlNode *child, xmlNode *update,
-                      bool delete_only)
+pcmk__xe_find_replace(xmlNode *xml, xmlNode *update, bool delete_only)
 {
-    /* @TODO Create a recursive helper so that callers don't need to pass a NULL
-     * parent argument.
-     *
-     * @TODO Try to extract delete into a different function from replace.
-     * Ideally separate tree traversal from any work.
-     *
-     * @COMPAT Is pcmk__apply_acl() enough to properly consider ACLs?
+    /* @COMPAT Is pcmk__apply_acl() enough to properly consider ACLs?
      *
      * @COMPAT Some of this behavior is questionable for general use but is
      * required for backward compatibility by cib_process_replace() and
      * cib_process_delete(). Consider moving some of this to libcib since those
      * are the only callers. Behavior can change at a major version release if
      * desired.
+     *
+     * @TODO Why don't we allow matching (and replacing or deleting) the tree
+     * root?
```

Comment:
I don't think so. Looking at `replace_xml_child()` on main...
* If `parent == NULL`, that means `child` is the tree root (top-level call).
* We check whether `child` is a match.
* If so, then if `parent != NULL` (recursive call), we delete/replace the child. Otherwise (if `parent == NULL, top-level call), we say "Cannot delete the search root" and then search all the children recursively.

So if the top-level call matches the tree root, we ignore it and search all its children.

Reply:
> It must work somehow though. I'm pretty sure it's possible to replace the entire CIB with the API call (via cibadmin).

There's a huge special case block for that in `cib_process_replace()` :)

```
    if (pcmk__xe_is(input, PCMK_XE_CIB)) {
    ... about 70 lines ...
    } else {
        xmlNode *obj_root = NULL;
        gboolean ok = TRUE;

        obj_root = pcmk_find_cib_element(*result_cib, section);
        ok = replace_xml_child(NULL, obj_root, input, FALSE);
        if (ok == FALSE) {
            crm_trace("No matching object to replace");
            result = -ENXIO;
        }
    }
```

Reply:
Ah, that would be why, then. They didn't want to accidentally skip that block. Probably a better way to do it, but not worth effort now